File: src/lib/playlistOrdering.ts
/**
 * Pure index math for reordering a playlist by one step, extracted so it's
 * testable without touching the database. Returns the pair of indices whose
 * sortOrder should be swapped, or null if the move is out of bounds.
 */
export function swapIndices(count: number, index: number, direction: -1 | 1): [number, number] | null {
  if (index < 0 || index >= count) return null;
  const target = index + direction;
  if (target < 0 || target >= count) return null;
  return [index, target];
}

File: Sources/App/Services/PlaylistOrdering.swift
/// Pure index arithmetic for the playlist Up/Down reorder buttons, kept
/// separate from `PlaylistController` so it's unit-testable without a
/// database or HTTP round-trip.
enum PlaylistOrdering {
    /// Given `count` entries (already sorted by position) and a 0-based
    /// `index` to move by `direction` (-1 = up, +1 = down), returns the pair
    /// of positions whose sort order should be swapped, or nil if the move
    /// would go out of bounds.
    static func swapIndices(count: Int, index: Int, direction: Int) -> (Int, Int)? {
        let swapIndex = index + direction
        guard (0..<count).contains(index), (0..<count).contains(swapIndex) else {
            return nil
        }
        return (index, swapIndex)
    }
}
